### PR TITLE
Add support for torch.gather (frontend)

### PR DESCRIPTION
**Pull request by ftxj**
This PR is frontend implementation of `torch.gather(lookup_tv, dim, index_tv, *, sparse_grad = False)` fusion

### CODE REVIEW

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 1]
```diff
@@ -222,6 +222,17 @@ const std::vector<std::string> functions = {
...
             return torch.zeros(sizes).to(grad).scatter_(dim, indices, grad)
 
+        def gather(self,
+                    dim: int,
+                    index,
+                    *,
+                    sparse_grad: bool = False):
+            output = torch.gather(self, dim, index)
+            def backward(grad_output):
+                grad_self = torch.zeros_like(self).scatter_add_(dim, index, grad_output)
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp]: Let's not use in-place operations unnecessarily. `scatter_add_` -> `scatter_add`

Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp]: nitpick: we are dropping `sparse_grad` on backwards, which doesn't seem to be right. Let's at least add a conditional here and go through `torch.gather_backward` when `sparse_grad` is True.

I'm hoping that `sparse_grad` will be constant most times and the conditional block can be constant folded, so we'll still be able to fuse the backward ops.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -4365,6 +4415,22 @@ bool insertProfileIValue(ProfilingRecord* pr, Node* node, size_t offset) {
     return true;
   }
 
+  static auto torch_gather_schema =
+      getOperatorForLiteral(
+          "aten::gather(Tensor self, int dim, Tensor index, *, bool sparse_grad=False) -> Tensor")
+          ->schema();
+  if (node->matches(torch_gather_schema)) {
+    switch (offset) {
+      // argument 1: unsqueeze dim;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nitpick: `unsqueeze dim` -> `gather dim`

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -1560,7 +1560,57 @@ class IrParser {
...
+      if (isOptionEnabled(EnableOption::GraphOp)) {
+        auto ptr_op = getOperatorForLiteral(
+            "aten::gather(Tensor self, int dim, Tensor index, *, bool sparse_grad=False) -> Tensor");
+        REGISTER_PARSE_RULE(
+            ptr_op,
+            {
+              MemoryFormat format;
+              std::list<Val*> list_val;
+              std::tie(format, list_val) = getPWFormatValues(
+                  c10::nullopt,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: I don't think the parsing rule is permutation agnostic. Let's change this from `c10::nullopt` -> `MemoryFormat::Contiguous()`

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Noticed that this is also the case for `index_select`, which I missed in my last review. Can we also update that. :stuck_out_tongue_winking_eye:

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -1560,7 +1560,57 @@ class IrParser {
...
+                  value_map[node->inputs()[0]->unique()],
+                  value_map[node->inputs()[2]->unique()]);
+              auto input = list_val.front();
+              list_val.pop_front();
+              auto dim_value = constant_as<int>(node->input(1));
+              TORCH_INTERNAL_ASSERT(
+                  dim_value.has_value(), "dim parameter is required.");
+              auto sparse_grad = constant_as<bool>(node->input(3));
+              TORCH_INTERNAL_ASSERT(
+                  sparse_grad.value() == false,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: I thought `sparse_grad` only affects backward ops?! Do we need to special case it here?!

File: test/test_jit_cuda_fuser.py [hunk 5]
```diff
@@ -4233,6 +4233,24 @@ def t(x_kj: torch.Tensor, idx_kj: torch.Tensor, sbf: torch.Tensor, dim : int):
...
+        lookup_tv = torch.rand(lookup_size, feat_dim, dtype=torch.float, device="cuda")
+        indies_tv = torch.randint(0, lookup_size, (lookup_size, select_dim), device="cuda").to(dtype=torch.long)
+        sbf = torch.rand(lookup_size, select_dim, dtype=torch.float, device="cuda")
+
+        def t(x_kj, idx_kj, inp):
+            gather_res = torch.gather(x_kj, 1, idx_kj) * inp
+            res = gather_res + 17
+            return res
+        t_jit = torch.jit.script(t)
+        self._run_helper(t_jit, t, lookup_tv, indies_tv, sbf)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: We are not testing autodiff here. Can we try with `_run_training_helper`?

I would also suggest us testing `sparse_grad` here as well (with it being false & true).

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 6]
```diff
@@ -1365,6 +1365,28 @@ void initNvFuserPythonBindings(PyObject* module) {
       py::arg("dim"),
       py::arg("index"),
       py::return_value_policy::reference);
+  nvf_ops.def(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp]: Looks about right to me. But the python API is not tested. Can we also add tests for this?

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 1 above)
Comment (ftxj) [reply torch/csrc/jit/runtime/symbolic_script.cpp]: Ok, but I don't know how to call `torch.gather_backward` here. When `sparse_grad` is True, the actually executed function is `at::_gather_sparse_backward`, but it does not provide python interface (maybe just my mistake). Maybe I need reimplement it in python?

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 7]
```diff
@@ -221,6 +221,52 @@ const std::vector<std::string> functions = {
...
+                for i in range(self.dim()) :
+                    n_above = int(n_above / grad.size(i))
+                    if (i == dim) :
+                        sparse_ind[i] = index.reshape(-1)
+                    else: 
+                        sparse_ind[i] = torch.arange(grad.size(i)).unsqueeze(1).expand(grad.size(i), n_above).reshape(-1).repeat(n_below).to(index)
+                    
+                    n_below *= grad.size(i)
+            return torch._sparse_coo_tensor_unsafe(sparse_ind, grad.reshape(-1), self.size())
+
```
Comment (ftxj) [inline torch/csrc/jit/runtime/symbolic_script.cpp:256]: `AD_gather_sparse_backend` is just a python implementation of `at::_gather_sparse_backward`, which is the grad function of `torch.gather` when `sparse_grad` is `True`.

File: test/test_jit_cuda_fuser.py [hunk 8]
```diff
@@ -4233,6 +4233,59 @@ def t(x_kj: torch.Tensor, idx_kj: torch.Tensor, sbf: torch.Tensor, dim : int):
...
+        y = torch.randint(0, 1, (2, 2), device="cuda").to(dtype=torch.long)
+        z = torch.rand(2, 2, dtype=torch.float, device="cuda").requires_grad_()
+        grad = torch.randn(2, 2, dtype=torch.float, device="cuda")
+        def t(x, y, z):
+            o = torch.gather(x + z, 1, y + y, sparse_grad=True)
+            return o
+
+        t_jit = torch.jit.script(t)
+        self._run_training_helper(t_jit, t, grad, x, y, z)
+
```
Comment (ftxj) [inline test/test_jit_cuda_fuser.py:4278]: Actually, `test_gather_sparse_grad` will raise `graph contains 0 prim::CudaFusionGuard nodes (including subgraphs) but expected 1` error. But I do not know how to fix it.

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp (hunk 6 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp]: Sorry, I do not know what to test here and where is test file for `nvf_ops.def`.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 1 above)
Comment (ftxj) [reply torch/csrc/jit/runtime/symbolic_script.cpp]: Done in new commit

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (ftxj) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Done in new commit

File: test/test_jit_cuda_fuser.py (hunk 5 above)
Comment (ftxj) [reply test/test_jit_cuda_fuser.py]: Done in new commit.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/symbolic_script.cpp]: Sorry I didn't respond in time for this. I see you already did your python version here.

Does this not work? vvv
`
if (sparse_grad):
    return torch.gather_backward(grad_output, self, dim, True), None, None, None
`

Just FYI:
1. You can check out aten signature here: https://github.com/pytorch/pytorch/blob/dabf515c184c18eb5a7ec2562110f957eb899427/aten/src/ATen/native/native_functions.yaml#L8556-L8559
2. autograd rule can be looked up here: https://github.com/pytorch/pytorch/blob/dabf515c184c18eb5a7ec2562110f957eb899427/tools/autograd/derivatives.yaml#L687-L690

File: test/test_jit_cuda_fuser.py (hunk 8 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:4278]: It should be failing here: https://github.com/csarofeen/pytorch/blob/1f39c3a1cacd1e83ab7dd519931076567fe81626/test/test_jit_cuda_fuser.py#L252

Since we don't have a backward rule for gather yet.
We can add two more default arguments to the function. `num_fw_fusion=1, num_bw_fusion=1` there, similar to what we did with `run_helper`: https://github.com/csarofeen/pytorch/blob/1f39c3a1cacd1e83ab7dd519931076567fe81626/test/test_jit_cuda_fuser.py#L197

Thanks in advance for doing the clean up :stuck_out_tongue_winking_eye:

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 9]
```diff
@@ -221,6 +221,52 @@ const std::vector<std::string> functions = {
 
             # FIXME: torchscript: torch.zeros(sizes, grad.options())
             return torch.zeros(sizes).to(grad).scatter_(dim, indices, grad)
+        
+        # python implementation of at::_gather_sparse_backward function. Need to Refine
+        def AD_gather_sparse_backend(grad,
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp]: this looks identical to what's in `_gather_sparse_backward`. Is there any good reason that we are re-doing this in python?

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 1 above)
Comment (ftxj) [reply torch/csrc/jit/runtime/symbolic_script.cpp]: I think `gather_backward` is not registered in torch, it just an `aten` function.

File: torch/csrc/jit/runtime/symbolic_script.cpp (hunk 9 above)
Comment (ftxj) [reply torch/csrc/jit/runtime/symbolic_script.cpp]: Just because of `gather_backward` is an `aten` function, not `torch`.

**Comment by jjsjann123** [comment]
Is this one ready for review again?

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 10]
```diff
@@ -1369,8 +1369,30 @@ void initNvFuserPythonBindings(PyObject* module) {
...
       py::arg("arg"),
-      py::arg("dim"),
       py::arg("index"),
+      py::arg("dim"),
+      py::return_value_policy::reference);
+  nvf_ops.def(
+      "gather",
+      [](nvfuser::FusionDefinition::Operators& self,
+         nvfuser::Tensor arg1,
+         nvfuser::Tensor arg3,
```
Comment (jjsjann123) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:1379]: nitpick: `arg3` -> `arg2`

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 11]
```diff
@@ -1369,8 +1369,30 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+            {
+                fd->recordingState(arg1()),
+                fd->recordingState(arg3()),
+            },
+            {fd->recordingState(output())},
+            dim));
+        return output;
+      },
+      py::arg("arg1"),
+      py::arg("arg3"),
```
Comment (jjsjann123) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:1394]: ditto

**Review by jjsjann123 — APPROVED** [review]
LGTM.

Thanks for going through this long review process :stuck_out_tongue_winking_eye:

**Comment by jjsjann123** [comment]
Realized that @ftxj can't merge PRs. I'll try land this myself after running through tests.

**Comment by jjsjann123** [comment]
Local tests passed. I'm merging this.

cc'ing @ftxj
